File: language-support/js/daml-ledger/src/mapping/tree_event.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as mapping from '.';
import * as ledger from '..';
import * as grpc from 'daml-grpc';

export const TreeEvent: mapping.Mapping<grpc.TreeEvent, ledger.TreeEvent> = {
    toObject(message: grpc.TreeEvent): ledger.TreeEvent {
        const object: ledger.TreeEvent = {};
        if (message.hasCreated()) {
            object.created = mapping.CreatedEvent.toObject(message.getCreated()!);
        }
        if (message.hasExercised()) {
            object.exercised = mapping.ExercisedEvent.toObject(message.getExercised()!);
        }
        return object;
    },
    toMessage(object: ledger.TreeEvent): grpc.TreeEvent {
        const message = new grpc.TreeEvent();
        if (object.created) {
            message.setCreated(mapping.CreatedEvent.toMessage(object.created));
        }
        if (object.exercised) {
            message.setExercised(mapping.ExercisedEvent.toMessage(object.exercised));
        }
        return message;
    }
}

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/TreeEvent.java
// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v1.TransactionOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.List;

/**
 * This interface represents events in {@link TransactionTree}s.
 *
 * @see CreatedEvent
 * @see ExercisedEvent
 * @see TransactionTree
 */
public interface TreeEvent {

    @NonNull List<@NonNull String> getWitnessParties();

    @NonNull String getEventId();

    @NonNull Identifier getTemplateId();

    @NonNull String getContractId();

    default TransactionOuterClass.TreeEvent toProtoTreeEvent() {
        TransactionOuterClass.TreeEvent.Builder eventBuilder = TransactionOuterClass.TreeEvent.newBuilder();
        if (this instanceof CreatedEvent) {
            CreatedEvent event = (CreatedEvent) this;
            eventBuilder.setCreated(event.toProto());
        } else if (this instanceof ExercisedEvent) {
            ExercisedEvent event = (ExercisedEvent) this;
            eventBuilder.setExercised(event.toProto());
        } else {
            throw new RuntimeException("this should be CreatedEvent or ExercisedEvent, found " + this.toString());
        }
        return eventBuilder.build();
    }

    static TreeEvent fromProtoTreeEvent(TransactionOuterClass.TreeEvent event) {
        if (event.hasCreated()) {
            return CreatedEvent.fromProto(event.getCreated());
        } else if (event.hasExercised()) {
            return ExercisedEvent.fromProto(event.getExercised());
        } else {
            throw new UnsupportedEventTypeException(event.toString());
        }
    }
}

